config.org
#+TITLE: DOOM Emacs Configuration
#+AUTHOR: Edmund Miller
* Basics
Who are you?
Who, who, who, who?
#+BEGIN_SRC emacs-lisp
(setq user-mail-address "Edmund.A.Miller@gmail.com"
      user-full-name "Edmund Miller")
#+END_SRC
#+BEGIN_SRC emacs-lisp
(defvar xdg-data (getenv "XDG_DATA_HOME"))
(defvar xdg-config (getenv "XDG_CONFIG_HOME"))
#+END_SRC
Change font
#+BEGIN_SRC emacs-lisp
(setq doom-font (font-spec :family "monospace" :size 15 :weight 'semi-light)
      doom-variable-pitch-font (font-spec :family "sans" :size 16))
#+END_SRC
Change Theme
#+BEGIN_SRC emacs-lisp
(setq doom-theme 'doom-flatwhite)
#+END_SRC
Change the banner
#+BEGIN_SRC emacs-lisp
(setq +doom-dashboard-banner-file (expand-file-name "narf.png" doom-private-dir))
#+END_SRC
Cleanup Whitespace
#+BEGIN_SRC emacs-lisp
;; (add-hook 'before-save-hook #'delete-trailing-whitespace)
#+END_SRC
#+begin_quote
Line numbers are pretty slow all around. The performance boost of
disabling them outweighs the utility of always keeping them on.
#+end_quote
#+BEGIN_SRC emacs-lisp
(setq display-line-numbers-type nil)
;; Prevents some cases of Emacs flickering
(add-to-list 'default-frame-alist '(inhibit-double-buffering . t))
#+END_SRC
* Keybindings
#+BEGIN_SRC emacs-lisp
(map!
 (:leader
  (:prefix "f"
   :desc "Find file in dotfiles" :n "o" #'+emiller/find-in-dotfiles
   :desc "Browse dotfiles" :n "O" #'+emiller/browse-dotfiles)
  (:prefix "i"
   :desc "Insert date" :n "d" #'insert-todays-date)
  (:prefix "n"
   :desc "Browse mode notes"     :n  "m" #'+emiller/find-notes-for-major-mode
   :desc "Browse project notes" :n "p" #'+emiller/find-notes-for-project)
  (:prefix "o"
   :desc "Calc" :n "c" #'calc
   :desc "easy-hugo" :n "g" #'easy-hugo
   :desc "APP: IRC" :n "i" #'=irc
   ;; :desc "APP: notmuch" :n "m" #'=mu4e
   :desc "dired-sidebar" :n "n" #'dired-sidebar-toggle-sidebar
   :desc "todo.org" :n "o" #'+emiller/visit-todo-org
   :desc "projects" :n "p" #'+emiller/visit-projects-org
   :desc "emms" :n "s" #'emms
   :desc "APP: rss" :n "," #'=rss)))
#+END_SRC
* Functions
Ediff the ~init.example.el~ and my ~init.el~
** Compare init files
#+BEGIN_SRC emacs-lisp
(defun ediff-init-files ()
  (interactive)
  (ediff-files (expand-file-name "init.el" doom-private-dir)
               (expand-file-name "init.example.el" user-emacs-directory)))
#+END_SRC
** Insert Date
Helpful for ~Snakemake~
#+BEGIN_SRC emacs-lisp
(defun insert-todays-date (arg)
  (interactive "P")
  (insert (if arg
              (format-time-string "%d-%m-%Y")
            (format-time-string "%Y-%m-%d"))))
(global-set-key (kbd "C-c d") 'insert-todays-date)
#+END_SRC
* Modules
** biblio
#+BEGIN_SRC emacs-lisp
(setq! +biblio-pdf-library-dir "~/sync/papers"
       +biblio-default-bibliography-files '("~/sync/reference/bibliography.bib")
       +biblio-notes-path "~/sync/org/roam/bib/")
#+END_SRC
** clojure
#+BEGIN_SRC emacs-lisp
(after! cider
  (set-popup-rules!
    '(("^\\*cider-error*" :ignore t)
      ("^\\*cider-repl" :quit nil)
      ("^\\*cider-repl-history" :vslot 2 :ttl nil))))
#+END_SRC
** company
This is no longer the default
#+BEGIN_SRC emacs-lisp
(setq company-idle-delay nil)
#+END_SRC
** dired
Get rid of dired message when using a
#+BEGIN_SRC emacs-lisp
(put 'dired-find-alternate-file 'disabled nil)
#+END_SRC
*** dired-sidebar
#+BEGIN_SRC emacs-lisp
(after! dired
  (use-package! dired-sidebar))
#+END_SRC
** direnv
#+BEGIN_SRC emacs-lisp
(setq direnv-always-show-summary nil)
#+END_SRC
** elisp
*** package-lint
#+BEGIN_SRC emacs-lisp
;; (use-package! package-lint)
#+END_SRC
** eshell
Get rid of modeline in eshell buffers
#+BEGIN_SRC emacs-lisp
(add-hook 'eshell-mode-hook #'hide-mode-line-mode)
#+END_SRC
** graphviz-dot-mode
#+BEGIN_SRC emacs-lisp
(use-package! graphviz-dot-mode)
#+END_SRC
** irc
#+BEGIN_SRC emacs-lisp
(after! circe
  (set-irc-server! "chat.freenode.net"
                   `(:tls t
                          :port 6697
                          :nick "emiller88"
                          :sasl-username ,(+pass-get-user "irc/freenode.net")
                          :sasl-password (lambda (&rest _) (+pass-get-secret "irc/freenode.net"))
                          :channels ("#bioinformatics" "#clojure" "#emacs" "#emacs-circe" "#guix" "#guile" "#home-manager" "#nixos" "#nixos-emacs" "#sway" "##rust" "#python" "#pine64"))))
#+END_SRC
Other Channels
- #haskell
** javascript
#+BEGIN_SRC emacs-lisp
(defun i-will-have-what-js-is-having ()
  (let ((buffer-file-name (concat (file-name-sans-extension buffer-file-name) ".js")))
    (set-auto-mode)))
(add-to-list 'auto-mode-alist '("\\.vmjs\\'" . i-will-have-what-js-is-having))
#+END_SRC
** ledger
#+BEGIN_SRC emacs-lisp
(add-to-list 'auto-mode-alist '("\\.\\(h?ledger\\|journal\\|j\\)$" . ledger-mode))
#+END_SRC
** lsp
#+BEGIN_SRC emacs-lisp
(after! lsp
  (setq lsp-ui-sideline-show-code-actions nil
        lsp-ui-sideline-show-diagnostics nil
        lsp-signature-render-all nil))
#+END_SRC
** magit
*** Git-gutter-fringe
#+BEGIN_SRC emacs-lisp
(defun +version-control|git-gutter-maybe ()
  (when buffer-file-name
    (require 'git-gutter-fringe)
    (git-gutter-mode +1)))
#+END_SRC
*** GPG signing
#+BEGIN_SRC emacs-lisp
(setq magit-repository-directories '(("~/src" . 3))
      magit-save-repository-buffers nil
      magit-inhibit-save-previous-winconf t
      transient-values '((magit-commit "--gpg-sign=BD387FF7BC10AA9D")
                         (magit-rebase "--autosquash" "--gpg-sign=BD387FF7BC10AA9D")
                         (magit-pull "--rebase" "--gpg-sign=BD387FF7BC10AA9D")))
#+END_SRC
*** Set editor
#+BEGIN_SRC emacs-lisp
(setenv "EDITOR" "emacsclient")
#+END_SRC
** mu4e
*** Setting mail dir
#+BEGIN_SRC emacs-lisp
;; Each path is relative to `+mu4e-mu4e-mail-path', which is ~/.mail by default
(after! mu4e
  (setq shr-use-colors nil)
  (set-email-account! "Gmail"
                      '((mu4e-sent-folder       . "/Gmail/[Gmail]/Sent Mail")
                        (mu4e-drafts-folder     . "/Gmail/[Gmail]/Drafts")
                        (mu4e-trash-folder      . "/Gmail/[Gmail]/Trash")
                        (mu4e-refile-folder     . "/Gmail/[Gmail]/Archive")
                        (smtpmail-smtp-user     . "edmund.a.miller@gmail.com")
                        (mu4e-compose-signature . "---\nEdmund Miller"))
                      t)

  (set-email-account! "Eman"
                      '((mu4e-sent-folder       . "/Eman/[Gmail]/Sent Mail")
                        (mu4e-drafts-folder     . "/Eman/[Gmail]/Drafts")
                        (mu4e-trash-folder      . "/Eman/[Gmail]/Trash")
                        (mu4e-refile-folder     . "/Eman/[Gmail]/Archive")
                        (mu4e-compose-signature . "---\nEdmund Miller"))
                      t)

  (set-email-account! "UTD"
                      '((mu4e-sent-folder       . "/UTD/Sent Items")
                        (mu4e-drafts-folder     . "/UTD/Drafts")
                        (mu4e-trash-folder      . "/UTD/Deleted Items")
                        (mu4e-refile-folder     . "/UTD/Archive")
                        (smtpmail-smtp-user     . "Edmund.Miller@utdallas.edu")
                        (mu4e-compose-signature . "---\nEdmund Miller"))
                      t))
#+END_SRC
*** mu4e-conversation
#+BEGIN_SRC emacs-lisp
;; (use-package! mu4e-conversation
;;   :after mu4e
;;   :init
;;   (global-mu4e-conversation-mode))
#+END_SRC
*** mu4e-patch
colorize patch-based emails
#+BEGIN_SRC emacs-lisp
(use-package! mu4e-patch
  :hook (mu4e-view-mode . mu4e-patch-highlight))
#+END_SRC
** nextflow
#+BEGIN_SRC emacs-lisp
(use-package! nextflow-mode)
#+END_SRC
** org
*** faces
I like to cross things off my todo list
#+BEGIN_SRC emacs-lisp
(custom-set-faces! '(org-headline-done :strike-through t))
#+END_SRC
*** Misc
Start in insert mode in =org-capture=
#+BEGIN_SRC emacs-lisp
(add-hook 'org-capture-mode-hook 'evil-insert-state)
#+END_SRC
Set ~+org-vars~
#+BEGIN_SRC emacs-lisp
(setq org-directory "/home/emiller/sync/org/"
      org-archive-location (concat org-directory "archive/%s::")
      org-ellipsis " ▼ "
      org-superstar-headline-bullets-list '("#"))
#+END_SRC
Org export settings
#+BEGIN_SRC emacs-lisp
(after! org
  (setq org-export-with-toc nil))
#+END_SRC
Log time when things get marked as done
#+BEGIN_SRC emacs-lisp
(after! org
  (setq org-log-done 'time))
#+END_SRC
*** Org Capture Templates
#+BEGIN_SRC emacs-lisp
(after! org
  (setq org-capture-templates
        (append
         ;; TODO generalize these with org-directory
         '(("a" "Appointment" entry (file  "~/sync/org/schedule.org" )
            "* %?\n\n%^T\n\n:PROPERTIES:\n\n:END:\n\n")
           ("l" "Lab Entry" entry
            (file+olp+datetree "~/sync/org/roam/lab/notebook.org")
            "* %U %?\n%i\n%a"))
         org-capture-templates)))
#+END_SRC
*** org-agenda-custom
#+BEGIN_SRC emacs-lisp
(after! org
  (setq org-agenda-custom-commands
        '(("c" "Simple agenda view"
           ((tags "PRIORITY=\"A\""
                  ((org-agenda-skip-function '(org-agenda-skip-entry-if 'todo 'done))
                   (org-agenda-overriding-header "High-priority unfinished tasks:")))
            (agenda "" nil)
            (alltodo "")))
          ("g" . "GTD contexts")
          ("ga" "All TODO" alltodo nil
           ((org-agenda-sorting-strategy '(tag-up priority-up))))

          ("gl" "Lab" tags-todo "lab"
           ((org-agenda-sorting-strategy '(priority-up))
            ;; (org-agenda-prefix-format "[ ] %T: ")
            (org-agenda-compact-blocks t)))

          ("gw" "Work" tags-todo "@work"
           ((org-agenda-sorting-strategy '(priority-up))))
          ;; (org-agenda-compact-blocks t)))

          ("gs" "Study Time" tags-todo "@study"
           ((org-agenda-sorting-strategy '(priority-up))))

          ("ge" "Emacs Time" tags-todo "@emacs")
          ("o" "My Agenda"
            (
           ;; ((todo "TODO" ()
           ;;        (org-agenda-overriding-header "\n⚡ Do Today:\n⎺⎺⎺⎺⎺⎺⎺⎺⎺")
           ;;        (org-agenda-remove-tags t)
           ;;        (org-agenda-prefix-format " %-2i %-15b")
           ;;        (org-agenda-todo-keyword-format ""))

            (agenda "" (
                        (org-agenda-start-day "+0d")
                        (org-agenda-span 5)
                        (org-agenda-overriding-header "⚡ Schedule:\n⎺⎺⎺⎺⎺⎺⎺⎺⎺")
                        (org-agenda-repeating-timestamp-show-all nil)
                        (org-agenda-remove-tags t)
                        (org-agenda-prefix-format   "  %-3i  %-15b %t%s")
                        (org-agenda-todo-keyword-format " ☐ ")
                        (org-agenda-current-time-string "⮜┈┈┈┈┈┈┈ now")
                        (org-agenda-scheduled-leaders '("" ""))
                        (org-agenda-time-grid (quote ((daily today remove-match)
                                                      (0900 1200 1500 1800 2100)
                                                      "      " "┈┈┈┈┈┈┈┈┈┈┈┈┈"))))))))))
#+END_SRC
*** ~Packages~
**** Auto-org-md
#+BEGIN_SRC emacs-lisp
(after! org
  (use-package! auto-org-md))
#+END_SRC
**** deft
#+BEGIN_SRC emacs-lisp
(after! deft
  (setq deft-directory "~/sync/org/roam"
        deft-recursive t))
#+END_SRC
**** Org-clock-csv
#+BEGIN_SRC emacs-lisp
(after! org
  (use-package! org-clock-csv))
#+END_SRC
**** org-gcal
#+BEGIN_SRC emacs-lisp
(use-package! org-gcal
  :config
  (setq org-gcal-client-id "119671856150-j6j4b8hjm1k8d1v2ar39c2g1ifdv8iq9.apps.googleusercontent.com"
        org-gcal-client-secret "KOa_aQ-SsyNkK_K4Y5ePk-k1"
        ;; TODO Generalize
        org-gcal-file-alist '(("Edmund.a.miller@gmail.com" .  "~/sync/org/schedule.org"))))
#+END_SRC
**** Org-noter
#+BEGIN_SRC emacs-lisp
(after! org-noter
  (setq org-noter-always-create-frame t
        org-noter-separate-notes-from-heading t
        org-noter-default-heading-title "Page $p$"
        org-noter-auto-save-last-location t
        org-noter-notes-search-path '("~/sync/org/roam/noter")
        org-noter-separate-notes-from-heading t
        org-noter-doc-property-in-notes t))
#+END_SRC
**** org-roam
#+BEGIN_SRC emacs-lisp
(after! org-roam
  (setq org-roam-capture-templates
        '(("l" "lit" plain (function org-roam--capture-get-point)
           "%?"
           :file-name "lit/${slug}"
           :head "#+setupfile:./hugo_setup.org
,#+hugo_slug: ${slug}
,#+title: ${title}\n"
           :unnarrowed t)
          ("c" "concept" plain (function org-roam--capture-get-point)
           "%?"
           :file-name "concepts/${slug}"
           :head "#+setupfile:./hugo_setup.org
,#+hugo_slug: ${slug}
,#+title: ${title}\n"
           :unnarrowed t)
          ("p" "private" plain (function org-roam-capture--get-point)
           "%?"
           :file-name "private/${slug}"
           :head "#+title: ${title}\n"
           :unnarrowed t)))
  (setq org-roam-capture-ref-templates
        '(("r" "ref" plain (function org-roam-capture--get-point)
           "%?"
           :file-name "lit/${slug}"
           :head "#+setupfile:./hugo_setup.org
,#+roam_key: ${ref}
,#+hugo_slug: ${slug}
,#+roam_tags: website
,#+title: ${title}
- source :: ${ref}"
           :unnarrowed t))))
#+END_SRC
**** journal
#+BEGIN_SRC emacs-lisp
(setq org-journal-dir "/home/emiller/sync/org/roam/journal/"
      org-journal-encrypt-journal t)
#+END_SRC
** python
#+BEGIN_SRC emacs-lisp
(after! lsp-python-ms
  (setq lsp-python-ms-executable (executable-find "python-language-server"))
  (set-lsp-priority! 'mspyls 1))
(set-eglot-client! 'python-mode (list (executable-find "python-language-server")))
#+END_SRC
** reference
#+BEGIN_SRC emacs-lisp
;; TODO Generalize
;; (setq reftex-default-bibliography '("~/sync/reference/references.bib")
;;       bibtex-completion-bibliography "~/sync/reference/references.bib" ;the major bibtex file
;;       bibtex-completion-library-path "~/sync/reference/pdf/" ;the directory to store pdfs
;;       bibtex-completion-notes-path "~/sync/reference/ref-notes.org" ;the note file for reference notes
;;       )
#+END_SRC
** rss
*** Keybindings
#+BEGIN_SRC emacs-lisp
(after! elfeed-search
  (map! :map elfeed-search-mode-map
        :localleader
        :n "m" #'my/elfeed-search-view-hydra/body
        :n "s" #'elfeed-toggle-star
        :n "r" #'elfeed-update))
#+END_SRC
*** Set max width
#+BEGIN_SRC emacs-lisp
(after! elfeed
  (setq elfeed-search-title-max-width 120))
#+END_SRC
*** Filter Hydra
**** Macro and function
#+BEGIN_SRC emacs-lisp
(use-package! pretty-hydra)
(cl-defmacro unpackaged/elfeed-search-view-hydra-define (name body views)
  "Define a pretty hydra named NAME with BODY and VIEWS.
VIEWS is a plist: in it, each property is a string which becomes
a column header in the hydra, and each value is a list of lists
in this format: (KEY COMPONENT &optional LABEL).

The KEY is a key sequence passed to `kbd', like \"s\" or \"S
TAB\".  The COMPONENT is an Elfeed filter component, which may
begin with \"+\" or \"=\", and in which spaces are automatically
escaped as required by Elfeed.  The LABEL, if present, is a
string displayed next to the KEY; if absent, COMPONENT is
displayed.

In the resulting hydra, when KEY is pressed, the COMPONENT is
toggled in `elfeed-search-filter'.  It is toggled between three
states: normal, inverse, and absent.  For example, the component
\"+tag\" cycles between three states in the filter: \"+tag\",
\"-tag\", and \"\".  The appropriate inverse prefix is used
according to the component's prefix (i.e. for \"=\", the inverse
is \"~\", and for \"\" (a plain regexp), \"!\" is used).

These special components may be used to read choices from the
Elfeed database with completion and toggle them:

  :complete-age   Completes and sets the age token.
  :complete-feed  Completes and toggles a feed token.
  :complete-tag   Completes and toggles a tag token.
  nil             Sets default filter.

A complete example:

  (unpackaged/elfeed-search-view-hydra-define my/elfeed-search-view-hydra
    (:foreign-keys warn)
    (\"Views\"
     ((\"@\" :complete-age \"Date\")
      (\"d\" nil))
     \"Status\"
     ((\"su\" \"+unread\"))
     \"Feed\"
     ((\"f TAB\" :complete-feed \"Choose\")
      (\"fE\" \"=Planet Emacslife\" \"Planet Emacslife\"))
     \"Tags\"
     ((\"t TAB\" :complete-tag \"Choose\")
      (\"te\" \"+Emacs\"))
     \"\"
     ((\"tn\" \"+news\"))))"
  (declare (indent defun))
  (cl-labels ((escape-spaces (string)
                             ;; Return STRING with spaces escaped with "\s-".  Necessary
                             ;; because Elfeed treats all literal spaces as separating tokens.
                             (replace-regexp-in-string (rx space) "\\s-" string t t)))
    (let* ((completion-fns
            (list (cons :complete-age
                        (lambda ()
                          (interactive)
                          (save-match-data
                            (let* ((date-regexp (rx (group (or bos blank) "@" (1+ digit) (1+ (not blank)))))
                                   (date-tag (when (string-match date-regexp elfeed-search-filter)
                                               (match-string 1 elfeed-search-filter))))
                              (elfeed-search-set-filter
                               (replace-regexp-in-string date-regexp (read-string "Date: " date-tag)
                                                         elfeed-search-filter t t))))))
                  (cons :complete-feed
                        '(concat "=" (replace-regexp-in-string
                                      (rx space) "\\s-"
                                      (->> (hash-table-values elfeed-db-feeds)
                                           (--map (elfeed-meta it :title))
                                           (completing-read "Feed: ")
                                           regexp-quote) t t)))
                  (cons :complete-tag
                        '(concat "+" (completing-read "Tag: " (elfeed-db-get-all-tags))))))
           (body (append '(:title elfeed-search-filter :color pink :hint t :quit-key "q")
                         body))
           (heads (cl-loop for (heading views) on views by #'cddr
                           collect heading
                           collect (cl-loop for (key component label) in views
                                            collect
                                            `(,key
                                              ,(cl-typecase component
                                                 ((and function (not null))
                                                  ;; I don't understand why nil matches
                                                  ;; (or lambda function), but it does,
                                                  ;; so we have to account for it.  See
                                                  ;; (info-lookup-symbol 'cl-typep).
                                                  `(funcall ,component))
                                                 (string
                                                  `(elfeed-search-set-filter
                                                    (unpackaged/elfeed-search-filter-toggle-component
                                                     elfeed-search-filter ,(escape-spaces component))))
                                                 (otherwise
                                                  `(elfeed-search-set-filter
                                                    ,(when component
                                                       `(unpackaged/elfeed-search-filter-toggle-component
                                                         elfeed-search-filter ,component)))))
                                              ,(or label component "Default"))))))
      ;; I am so glad I discovered `cl-sublis'.  I tried several variations of `cl-labels' and
      ;; `cl-macrolet' and `cl-symbol-macrolet', but this is the only way that has worked.
      (setf heads (cl-sublis completion-fns heads))
      `(pretty-hydra-define ,name ,body
                            ,heads))))

(cl-defun unpackaged/elfeed-search-filter-toggle-component (string component)
  "Return STRING (which should be `elfeed-search-filter') having toggled COMPONENT.
Tries to intelligently handle components based on their prefix:
+tag, =feed, regexp."
  (save-match-data

eshell/history
git clone git@github.com:ar1a/dotfiles.git
ls
rm -rf dotfiles/
ls
cd
cd src/dotfiles/
git clone git@github.com:ar1a/dotfiles.git ar1a


config.el
;;; config-new.el -*- lexical-binding: t; -*-

(setq user-mail-address "Edmund.A.Miller@gmail.com"
      user-full-name "Edmund Miller"

      doom-scratch-initial-major-mode 'lisp-interaction-mode
      doom-theme 'doom-flatwhite
      treemacs-width 32

      ;; Line numbers are pretty slow all around. The performance boost of
      ;; disabling them outweighs the utility of always keeping them on.
      display-line-numbers-type nil

      ;; This used to be off by default
      company-idle-delay nil

      ;; lsp-ui-sideline is redundant with eldoc and much more invasive, so
      ;; disable it by default.
      lsp-ui-sideline-enable nil
      lsp-enable-symbol-highlighting nil

      ;; More common use-case
      evil-ex-substitute-global t)

;;
;;; UI

;; "monospace" means use the system default. However, the default is usually two
;; points larger than I'd like, so I specify size 12 here.
(setq doom-font (font-spec :family "monospace" :size 26 :weight 'semi-light)
      doom-variable-pitch-font (font-spec :family "sans" :size 27))

;; Prevents some cases of Emacs flickering
(add-to-list 'default-frame-alist '(inhibit-double-buffering . t))

;;
;;; Keybinds

(map!
 :n [tab] (cmds! (and (featurep! :editor fold)
                      (save-excursion (end-of-line) (invisible-p (point))))
                 #'+fold/toggle
                 (fboundp 'evil-jump-item)
                 #'evil-jump-item)
 :v [tab] (cmds! (and (bound-and-true-p yas-minor-mode)
                      (or (eq evil-visual-selection 'line)
                          (not (memq (char-after) (list ?\( ?\[ ?\{ ?\} ?\] ?\))))))
                 #'yas-insert-snippet
                 (fboundp 'evil-jump-item)
                 #'evil-jump-item)
 (:leader
  (:prefix "f"
   :desc "Find file in dotfiles" :n "o" #'+emiller/find-in-dotfiles
   :desc "Browse dotfiles" :n "O" #'+emiller/browse-dotfiles)
  (:prefix "i"
   :desc "Insert date" :n "d" #'insert-todays-date)
  (:prefix "n"
   :desc "Browse mode notes"     :n  "m" #'+emiller/find-notes-for-major-mode
   :desc "Browse project notes" :n "p" #'+emiller/find-notes-for-project)
  (:prefix "o"
   :desc "Calc" :n "c" #'calc
   :desc "easy-hugo" :n "g" #'easy-hugo
   :desc "APP: IRC" :n "i" #'=irc
   ;; :desc "APP: notmuch" :n "m" #'=mu4e
   ;; :desc "dired-sidebar" :n "n" #'dired-sidebar-toggle-sidebar
   :desc "todo.org" :n "o" #'+emiller/visit-todo-org
   :desc "projects" :n "p" #'+emiller/visit-projects-org
   :desc "emms" :n "s" #'emms
   :desc "APP: rss" :n "," #'=rss)))

;;
;;; Modules

;;; :completion ivy
(after! ivy
  ;; I prefer search matching to be ordered; it's more precise
  (add-to-list 'ivy-re-builders-alist '(counsel-projectile-find-file . ivy--regex-plus)))

;; Switch to the new window after splitting
(setq evil-split-window-below t
      evil-vsplit-window-right t)

;;; :emacs dired
;;Get rid of dired message when using "a"
(put 'dired-find-alternate-file 'disabled nil)

;;; :ui doom-dashboard
;; NARF
(setq fancy-splash-image (concat doom-private-dir "narf.png"))
;; Don't need the menu; I know them all by heart
(remove-hook '+doom-dashboard-functions #'doom-dashboard-widget-shortmenu)


;;; :tools biblio
(setq! +biblio-pdf-library-dir "~/sync/papers"
       +biblio-default-bibliography-files '("~/sync/reference/bibliography.bib")
       +biblio-notes-path "~/sync/org/roam/bib/")

;;; :tools direnv
(setq direnv-always-show-summary nil)

;;; :tools magit
(setq magit-repository-directories '(("~/src" . 3))
      magit-save-repository-buffers nil
      ;; Don't restore the wconf after quitting magit, it's jarring
      magit-inhibit-save-previous-winconf t
      transient-values '((magit-commit "--gpg-sign=BD387FF7BC10AA9D")
                         (magit-rebase "--autosquash" "--gpg-sign=BD387FF7BC10AA9D")
                         (magit-pull "--rebase" "--gpg-sign=BD387FF7BC10AA9D")))

;; Enable git gutter on tramp sessions
(defun +version-control|git-gutter-maybe ()
  (when buffer-file-name
    (require 'git-gutter-fringe)
    (git-gutter-mode +1)))

(setenv "EDITOR" "emacsclient")

;;; :lang clojure
(after! cider
  (set-popup-rules!
    '(("^\\*cider-error*" :ignore t)
      ("^\\*cider-repl" :quit nil)
      ("^\\*cider-repl-history" :vslot 2 :ttl nil))))

;;; :lang ledger
(add-to-list 'auto-mode-alist '("\\.\\(h?ledger\\|journal\\|j\\)$" . ledger-mode))

;;; :lang org
(setq org-directory "/home/emiller/sync/org/"
      org-archive-location (concat org-directory "archive/%s::")
      org-ellipsis " ▼ "
      org-superstar-headline-bullets-list '("#")
      org-export-with-toc nil
      org-log-done 'time)

(after! org
  (setq org-capture-templates
        (append
         ;; TODO generalize these with org-directory
         '(("a" "Appointment" entry (file  "~/sync/org/schedule.org")
            "* %?\n\n%^T\n\n:PROPERTIES:\n\n:END:\n\n"
            ("l" "Lab Entry" entry
             (file+olp+datetree "~/sync/org/roam/lab/notebook.org")
             "* %U %?\n%i\n%a")))
         org-capture-templates)))


;; I like to cross things off my todo list
(custom-set-faces! '(org-headline-done :strike-through t))

;; Start in insert mode in org-capture
(add-hook 'org-capture-mode-hook 'evil-insert-state)

;; +journal
(after! org
  (setq org-journal-dir (concat org-roam-directory "journal/")
        org-journal-encrypt-journal t
        org-journal-file-format "%Y%m%d.org"))

;; +noter
(after! org-noter
  (setq org-noter-always-create-frame t
        org-noter-doc-split-fraction '(0.65 . 0.35)
        org-noter-separate-notes-from-heading t
        org-noter-default-heading-title "Page $p$"
        org-noter-auto-save-last-location t
        org-noter-notes-search-path '("~/sync/org/roam/noter")
        org-noter-separate-notes-from-heading t
        org-noter-doc-property-in-notes t))

;; +roam
(after! org-roam
  (setq org-roam-capture-templates
        '(("l" "lit" plain (function org-roam--capture-get-point)
           "%?"
           :file-name "lit/${slug}"
           :head "#+setupfile:./hugo_setup.org
#+hugo_slug: ${slug}
#+title: ${title}\n"
           :unnarrowed t)
          ("c" "concept" plain (function org-roam--capture-get-point)
           "%?"
           :file-name "concepts/${slug}"
           :head "#+setupfile:./hugo_setup.org
#+hugo_slug: ${slug}
#+title: ${title}\n"
           :unnarrowed t)
          ("p" "private" plain (function org-roam-capture--get-point)
           "%?"
           :file-name "private/${slug}"
           :head "#+title: ${title}\n"
           :unnarrowed t)))
  (setq org-roam-capture-ref-templates
        '(("r" "ref" plain (function org-roam-capture--get-point)
           "%?"
           :file-name "lit/${slug}"
           :head "#+setupfile:./hugo_setup.org
#+roam_key: ${ref}
#+hugo_slug: ${slug}
#+roam_tags: website
#+title: ${title}

- source :: ${ref}"
           :unnarrowed t))))

(use-package! org-roam-server
  :ensure t
  :config
  (setq org-roam-server-host "127.0.0.1"
        org-roam-server-port 8080
        org-roam-server-authenticate nil
        org-roam-server-export-inline-images t
        org-roam-server-serve-files nil
        org-roam-server-served-file-extensions '("pdf" "mp4" "ogv")
        org-roam-server-network-poll t
        org-roam-server-network-arrows nil
        org-roam-server-network-label-truncate t
        org-roam-server-network-label-truncate-length 60
        org-roam-server-network-label-wrap-length 20))

;;; :lang python
(after! lsp-python-ms
  (setq lsp-python-ms-executable (executable-find "python-language-server"))
  (set-lsp-priority! 'mspyls 1))
(set-eglot-client! 'python-mode (list (executable-find "python-language-server")))

;;; :lang rust
(setq rustic-lsp-server 'rust-analyzer)

;;; :lang solidity
(setq flycheck-solidity-solium-soliumrcfile "/home/emiller/sync/.soliumrc.json"
      solidity-flycheck-solc-checker-active t)

;;; :lang web
;; Hook into vue-lsp
(add-hook 'web-mode-local-vars-hook #'lsp!)

;;; :email mu4e
;; Each path is relative to `+mu4e-mu4e-mail-path', which is ~/.mail by default
(after! mu4e
  (setq shr-use-colors nil)
  (set-email-account! "Gmail"
                      '((mu4e-sent-folder       . "/Gmail/[Gmail]/Sent Mail")
                        (mu4e-drafts-folder     . "/Gmail/[Gmail]/Drafts")
                        (mu4e-trash-folder      . "/Gmail/[Gmail]/Trash")
                        (mu4e-refile-folder     . "/Gmail/[Gmail]/Archive")
                        (smtpmail-smtp-user     . "edmund.a.miller@gmail.com")
                        (mu4e-compose-signature . "---\nEdmund Miller"))
                      t)

  (set-email-account! "Eman"
                      '((mu4e-sent-folder       . "/Eman/[Gmail]/Sent Mail")
                        (mu4e-drafts-folder     . "/Eman/[Gmail]/Drafts")
                        (mu4e-trash-folder      . "/Eman/[Gmail]/Trash")
                        (mu4e-refile-folder     . "/Eman/[Gmail]/Archive")
                        (mu4e-compose-signature . "---\nEdmund Miller"))
                      t)

  (set-email-account! "UTD"
                      '((mu4e-sent-folder       . "/UTD/Sent")
                        (mu4e-drafts-folder     . "/UTD/Drafts")
                        (mu4e-trash-folder      . "/UTD/Trash")
                        (mu4e-refile-folder     . "/UTD/Archive")
                        (smtpmail-smtp-user     . "Edmund.Miller@utdallas.edu")
                        (mu4e-compose-signature . "---\nEdmund Miller"))
                      t))

(use-package! mu4e-patch
  :hook (mu4e-view-mode . mu4e-patch-highlight))

;;; :app calendar
(use-package! org-gcal
  :config
  (setq org-gcal-client-id "119671856150-j6j4b8hjm1k8d1v2ar39c2g1ifdv8iq9.apps.googleusercontent.com"
        org-gcal-client-secret "KOa_aQ-SsyNkK_K4Y5ePk-k1"
        ;; TODO Generalize
        org-gcal-file-alist '(("Edmund.a.miller@gmail.com" .  "~/sync/org/schedule.org"))))

;;; :app irc
(after! circe
  (set-irc-server! "chat.freenode.net"
                   `(:tls t
                     :port 6697
                     :nick "emiller88"
                     :sasl-username ,(+pass-get-user "irc/freenode.net")
                     :sasl-password (lambda (&rest _) (+pass-get-secret "irc/freenode.net"))
                     :channels ("#bioinformatics" "#clojure" "#emacs" "#emacs-circe" "#guix" "#guile" "#home-manager" "#nixos" "#nixos-emacs" "#sway" "##rust" "#python" "#pine64"))))
;;; :app rss
(after! elfeed-search
  (map! :map elfeed-search-mode-map
        :localleader
        :n "m" #'my/elfeed-search-view-hydra/body
        :n "s" #'elfeed-toggle-star
        :n "r" #'elfeed-update))
;; Set max width
(after! elfeed
  (setq elfeed-search-title-max-width 120))


;;
;;; Language customizations

(custom-theme-set-faces! 'doom-dracula
  `(markdown-code-face :background ,(doom-darken 'bg 0.075))
  `(font-lock-variable-name-face :foreground ,(doom-lighten 'magenta 0.6)))


;;
;;; Packages

(use-package! graphviz-dot-mode)
(use-package! jest)
(use-package! org-clock-csv)
(use-package! nextflow-mode)
(after! org
  (use-package! org-clock-csv))

;;
;;; Custom Variables
(custom-set-variables
 ;; custom-set-variables was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(safe-local-variable-values
   '((ssh-deploy-async-with-threads . 1)
     (ssh-deploy-on-explicity-save . t)
     (ssh-deploy-async . 1))))
